walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,185 @@
+   * @return list of timeranges for availability.
+   */
+  public Collection<TimeRange> findPotentialTimes(List<Event> bookedEvents, MeetingRequest request){
+    // List of potential free timeranges for the meeting request.
+    List<TimeRange> potentialTime = new ArrayList<>();
+    
+    Event prevEvent = bookedEvents.get(0); 
+    Event currEvent = bookedEvents.get(0);
+    for(int i = 0; i < bookedEvents.size(); i++){ 
+      prevEvent = currEvent; 
+      currEvent = bookedEvents.get(i); 
+
+      // If only one event - consider only timerange before and after the event.
+      if(bookedEvents.size() == 1) {
+        potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(), false));
+        potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY, true));
+        break;
+      }
+
+      // No times found yet.
+      if(potentialTime.size() == 0) { 
+        // See if there is a slot before the first booked event and if this slot.
+        // is large enough for the request. 
+        if(currEvent.getWhen().start() != TimeRange.START_OF_DAY) {
+          // If request duration is within slot between start of day and first event (curr).
+          if((TimeRange.START_OF_DAY+request.getDuration()) < currEvent.getWhen().start()){
+            potentialTime.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, currEvent.getWhen().start(),false));  
+          }
+        }
+        // If meeting request can fit in between end of event and end of day.
+        if((currEvent.getWhen().end()+request.getDuration()) <= TimeRange.END_OF_DAY){
+            potentialTime.add(TimeRange.fromStartEnd(currEvent.getWhen().end(), TimeRange.END_OF_DAY,true));      
+        }
+      } else {
+        // Time slots have been found.
+        // Remove the last timerange from list (time is prevEvent.end to endOfDay) to check conflict with currEvent.
+        potentialTime.remove(potentialTime.size()-1);
+        
+        // Check to see if prev and curr events OVERLAP.
+        // Add timerange from the end of the later timerange.
+        if(prevEvent.getWhen().overlaps(currEvent.getWhen())){
+          int laterEnd = currEvent.getWhen().end();
+          if(prevEvent.getWhen().end() > laterEnd) {
+            laterEnd = prevEvent.getWhen().end();
+          }
+          potentialTime.add(TimeRange.fromStartEnd(laterEnd, TimeRange.END_OF_DAY,true));
+        } else {
+          // check GAP (...) in-between curr and prev.
+          // |--prev--|...|--curr--|
+          long gapDuration = currEvent.getWhen().start() - prevEvent.getWhen().end();
+          if(gapDuration >= request.getDuration()){
+            potentialTime.add(TimeRange.fromStartEnd(prevEvent.getWhen().end(), currEvent.getWhen().start(), false));
+          }
+          // Check there there is a gap between the end of the event and end of day large enough for request.
```

Comment:
If you compare an event against the previous event (which is always possible if you process the first event outside the loop, then there will always be a previous event) then assuming that the gap between them is big enough you can just add it to potentialTime without having to re-check it again later. This will allow you to remove a lot of logic, significantly reducing complexity.

Reply:
Done